File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    answer = 0
    
    N = len(maps)
    M = len(maps[0])
    
    visited = [[-1] * M for _ in range(N)]
    
    direct = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    q = deque()
    q.append((0, 0))
    visited[0][0] = 0
    
    while q:
        x, y = q.popleft()
        
        for dx, dy in direct:
            nx, ny = x + dx, y + dy
            
            if 0 <= nx < N and 0 <= ny < M and maps[nx][ny] == 1 and visited[nx][ny] == -1:
                visited[nx][ny] = visited[x][y] + 1
                q.append((nx, ny))
        
    if visited[N-1][M-1] == -1:
        return -1

    return visited[N-1][M-1] + 1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    var answer = 1;
    const N = maps.length;
    const M = maps[0].length;
    const direct = [[-1, 0], [0, 1], [1, 0], [0, -1]];
    let visited = maps;
    let queue = [];
    
    queue.push([0, 0]);
    visited[0][0] = 0;
    
    while (queue.length > 0) {        
        let queueLength = queue.length;
        for (let i = 0; i < queueLength; i++) {
            let [x, y] = queue.shift();
            
            for (let j = 0; j < 4; j++) {
                let nx = x + direct[j][0];
                let ny = y + direct[j][1];
                
                if (nx >= 0 && nx < N && ny >= 0 && ny < M && visited[nx][ny] === 1) {
                    if (nx == N-1 && ny == M-1) {
                        return ++answer;
                    }
                    queue.push([nx, ny]);
                    visited[nx][ny] = 0;
                }
            }
        }
        answer++;
    }
    
    return -1;
}